File: 814-binary-tree-pruning/814-binary-tree-pruning.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pruneTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def recSearch(root: Optional[TreeNode]):
            if (root.left is None) and (root.right is None):
                return (root.val == 1)
            res_left = False
            res_right = False
            if root.left:
                res_left = recSearch(root.left)
                if not(res_left):
                    root.left = None
            if root.right:
                res_right = recSearch(root.right)
                if not(res_right):
                    root.right = None
            return res_left or res_right or (root.val == 1)
        res = recSearch(root)
        if res:
            return root
        return None

File: 814-binary-tree-pruning/814-binary-tree-pruning.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool recSearch(TreeNode* root) {
        if ((root->left == NULL) and (root->right == NULL)) {
            return (root->val == 1);
        }
        bool res_left = false;
        bool res_right = false;
        if (root->left) {
            res_left = this->recSearch(root->left);
            if (not(res_left)) {
                root->left = NULL;
            }
        }
        if (root->right) {
            res_right = this->recSearch(root->right);
            if (not(res_right)) {
                root->right = NULL;
            }
        }
        return (res_left or res_right or (root->val == 1));
    }

    TreeNode* pruneTree(TreeNode* root) {
        bool res = this->recSearch(root);
        if (res) {
            return root;
        }
        return NULL;
    }
};